### ISSUE TITLE
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf

### ISSUE THREAD

**Issue by hanzigs** [microsoft/LightGBM#3559]
For a data with 24 columns and 1000 records, running automl with following parameter set
```
    "LGBM_Estimators" : [100,200,300,400,500,600,700,800,900,1000],	
    "LGBM_Learning_Rate" : [0.1,0.2,0.3,0.4,0.5,0.6,0.7,0.8,0.9,1.0],
    "LGBM_ColByTree" : [0.1,0.2,0.3,0.4,0.5,0.6,0.7,0.8,0.9,1.0],
    "LGBM_SubSample" : [0.1,0.2,0.3,0.4,0.5,0.6,0.7,0.8,0.9,1.0],
    "LGBM_MinCWeight" : [0.001,0.002,0.003,0.004,0.005,0.006,0.007,0.008,0.009],
    "LGBM_NumLeaves" : [8,16,24,32,36,40,46,52,58,64,68,72]
```
throws lot of repeated below warning
```
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
```

May i know whats the mistake i am making?
Thanks

**Comment by jameslamb** [comment]
Hi @hanzigs , thanks for using LightGBM!

This doesn't mean that you've made any "mistakes", necessarily. That warning means that the boosting process has effectively ended early, because you've overfit to the training data.

The default `min_data_in_leaf` for LightGBM is 20 (https://lightgbm.readthedocs.io/en/latest/Parameters.html#min_data_in_leaf). That means that any split that would create a leaf node with less than 20 records in it is ignored. Setting `num_leaves` to values as high as 72 will grow very deep trees, and I think it's too aggressive for a dataset with only 1000 records.

If you want to avoid this in the future, you can enable early stopping by passing `early_stopping_rounds` and a validation set: https://lightgbm.readthedocs.io/en/latest/Python-Intro.html?highlight=early%20stopping#early-stopping. That will stop the training process after a few consecutive iterations with no gain.

I also recommend not using `num_leaves` values larger than 32 with such a small dataset.

**Comment by hanzigs** [comment]
Thanks @jameslamb 
Its good now, for the automl I set the above default parameter set, so for testing, passed a small dataset and got the warnings

**Comment by hanzigs** [comment]
Hi @jameslamb 
I can understand this conversion "num_leaves = 2** max_depth" is not a good practice,
May i know for MaxDepth = [2,3,4,5,6,7,8,9,10]
What can be set for num_leaves? or can we leave as default=31
Thanks

**Comment by jameslamb** [comment]
Those hyperparameters (`num_leaves`, `max_depth`) need to be tuned, and good values will depend on your dataset.

When you think about which types of values to search, remember that LightGBM grows trees  leaf-wise, not level-wise:  https://lightgbm.readthedocs.io/en/latest/Features.html#leaf-wise-best-first-tree-growth.

You can use the combination of `min_data_in_leaf`, `max_depth`, and `num_leaves` to control complexity. 

If you want to make the problem simpler to reason about, you could keep  `max_depth = -1` (no limit on max depth) and only search over values of `num_leaves`.

Please see this document for more information: https://lightgbm.readthedocs.io/en/latest/Parameters-Tuning.html

**Comment by avm19** [comment]
> That warning means that the boosting process has effectively ended early, because you've overfit to the training data.
> ...
> If you want to avoid this in the future, you can enable early stopping by passing `early_stopping_rounds` and a validation set

It rather means that the tree growing process ended early (not the boosting process), right? It is important to distinguish the two. As far as I could see, this message never appears with stumps (`max_depth=1`). Also, it is possible (and likely) that _not fully grown_ trees improve the overall model, so that even with `early_stopping_rounds` enabled the training proceeds and early stopping is not triggered, as in my example:
```
[1]	validation_0's l1: 999.062
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[2]	validation_0's l1: 919.248
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[3]	validation_0's l1: 849.926
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[4]	validation_0's l1: 788.677
```

**Comment by jameslamb** [comment]
> It rather means that the tree growing process ended early (not the boosting process), right? 

You're right @avm19 . My language in https://github.com/microsoft/LightGBM/issues/3559#issuecomment-726170902 was imprecise, and I've just modified that comment.

This warning signals that the growth of *a single tree* has stopped before reaching the configured `max_depth` / `num_leaves`.

Boosting will continue until `num_iterations` trees have been built or early stopping conditions are met.
